openfe/setup/protein_molecule.py

```diff
@@ -0,0 +1,59 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+from rdkit import Chem
+
+from openfe.utils.typing import RDKitMol, OEMol
+
+
+class ProteinMolecule:
+    """Wrapper around a Protein representation
+
+    This representation is immutable.  If you want to make any modifications,
+    do this in an appropriate toolkit then remake this class.
+    """
+    def __init__(self, rdkit: RDKitMol, name=""):
+        self._rdkit = rdkit
+        self._name = name
+
+    @property
+    def name(self):
+        return self._name
+
+    @classmethod
+    def from_pdbfile(cls, pdbfile: str, name=""):
+        m = Chem.MolFromPDBFile(pdbfile)
+        if m is None:
+            raise ValueError(f"RDKit failed to produce a molecule from "
+                             "{pdbfile}")
+        return cls(rdkit=m, name=name)
+
+    @classmethod
+    def from_pdbxfile(cls, pdbxfile: str, name=""):
+        raise NotImplementedError()
+
+    @classmethod
+    def from_rdkit(cls, rdkit: RDKitMol, name=""):
+        return cls(rdkit=Chem.Mol(rdkit), name=name)
+
+    def to_rdkit(self) -> RDKitMol:
+        return Chem.Mol(self._rdkit)
+
+    @classmethod
+    def from_openff(cls, offmol, name=""):
+        raise NotImplementedError()
+
+    def to_openff(self):
+        raise NotImplementedError()
+
+    @classmethod
+    def from_openeye(cls, oemol: OEMol, name=""):
+        raise NotImplementedError()
+
+    def to_openeye(self) -> OEMol:
+        raise NotImplementedError()
+
+    def __hash__(self):
+        return hash((self.name, Chem.MolToSequence(self._rdkit)))
```

Comment:
As per one of my comments above, I'm not super versed on how RDKit deals with PDB files, does this hold up for noncanonical sequences and/or HETATM residues? (i.e. we should include these as part of the sequence)

Reply:
I think there's more to hashing here. We've already discussed protonation states. What about:

* crystallographic waters
* bound ions
* cofactors

To me, all of these play a conceptual role in the "protein", and we're going to get all of them from the PDB structures. (Otherwise, at what stage do we add them?)

Reply:
Assuming the hash is just a list of residues and pass it a sequence of residues that spans anything in the PDB, I _think_ that should all be accounted for right?

We do get the issue of coordinate clashes, but I think we just assumed that this was something we'd assume equality on unless folks pass a different name right?

Reply:
That would work in principle (if you take the 3-letter residue name, with "unofficial" protonation state names). That's not what the current implementation does, though; `MolToSequence` isn't the right thing.

A PDB file with that includes cofactor and a bunch of ions (some of which are actually solvent, one of which is bound):

```python
In [1]: from rdkit import Chem

In [2]: mol = Chem.MolFromPDBFile("kras.pdb")

In [3]: Chem.MolToSequence(mol)
Out[3]: 'MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPTIEDSYRKQVVIDGETCLLDILDTAGQEEYSAMRDQYMRTGEGFLCVFAINNTKSFEDIHHYREQIKRVKDSEDVPMVLVGNKCDLPSRTVDTKQAQDLARSYGIPFIETSAKTRQGVDDAFYTLVREIRKHX'

In [4]: len(_)
Out[4]: 167

In [5]: import mdtraj as md

In [6]: md.load("kras.pdb")
Out[6]: <mdtraj.Trajectory with 1 frames, 2722 atoms, 219 residues, and unitcells at 0x1119bf5e0>
```

Reply:
>  That's not what the current implementation does, though; MolToSequence isn't the right thing.

Yeah 100%, sorry I should have worded that a lot better, my previous comment was more of a "if we did that would that be sufficient or do we need something fancier".